pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/colocated/GreedyShuffleRewriteVisitor.java

```diff
@@ -0,0 +1,393 @@
+        colocationKeys = new HashSet<>();
+      }
+      context.setColocationKeys(node.getStageId(), colocationKeys);
+      return colocationKeys;
+    }
+    // If receiver is a join-stage, remember partition-keys of the child node of MailboxSendNode.
+    Set<ColocationKey> mailboxSendColocationKeys = canSkipShuffleBasic ? oldColocationKeys : new HashSet<>();
+    context.setColocationKeys(node.getStageId(), mailboxSendColocationKeys);
+    return mailboxSendColocationKeys;
+  }
+
+  @Override
+  public Set<ColocationKey> visitProject(ProjectNode node, GreedyShuffleRewriteContext context) {
+    // Project reorders or removes keys
+    Set<ColocationKey> oldColocationKeys = node.getInputs().get(0).visit(this, context);
+
+    Map<Integer, Integer> oldToNewIndex = new HashMap<>();
+    for (int i = 0; i < node.getProjects().size(); i++) {
+      RexExpression rex = node.getProjects().get(i);
+      if (rex instanceof RexExpression.InputRef) {
+        int index = ((RexExpression.InputRef) rex).getIndex();
+        oldToNewIndex.put(index, i);
+      }
+    }
+
+    return computeNewColocationKeys(oldColocationKeys, oldToNewIndex);
+  }
+
+  @Override
+  public Set<ColocationKey> visitSort(SortNode node, GreedyShuffleRewriteContext context) {
+    return node.getInputs().get(0).visit(this, context);
+  }
+
+  @Override
+  public Set<ColocationKey> visitTableScan(TableScanNode node, GreedyShuffleRewriteContext context) {
+    TableConfig tableConfig =
+        _tableCache.getTableConfig(node.getTableName());
+    if (tableConfig == null) {
+      LOGGER.warn("Couldn't find tableConfig for {}", node.getTableName());
+      return new HashSet<>();
+    }
+    IndexingConfig indexingConfig = tableConfig.getIndexingConfig();
+    if (indexingConfig != null && indexingConfig.getSegmentPartitionConfig() != null) {
+      Map<String, ColumnPartitionConfig> columnPartitionMap =
+          indexingConfig.getSegmentPartitionConfig().getColumnPartitionMap();
+      if (columnPartitionMap != null) {
+        Set<String> partitionColumns = columnPartitionMap.keySet();
+        Set<ColocationKey> newColocationKeys = new HashSet<>();
+        for (int i = 0; i < node.getTableScanColumns().size(); i++) {
+          String columnName = node.getTableScanColumns().get(i);
+          if (partitionColumns.contains(node.getTableScanColumns().get(i))) {
+            int numPartitions = columnPartitionMap.get(columnName).getNumPartitions();
+            String hashAlgorithm = columnPartitionMap.get(columnName).getFunctionName();
+            newColocationKeys.add(new ColocationKey(i, numPartitions, hashAlgorithm));
+          }
+        }
+        return newColocationKeys;
+      }
+    }
+    return new HashSet<>();
+  }
+
+  @Override
+  public Set<ColocationKey> visitValue(ValueNode node, GreedyShuffleRewriteContext context) {
+    return new HashSet<>();
+  }
+
+  // TODO: Only equality joins can be colocated. We don't have join clause info available right now.
```

Comment:
how do we plan to get the info?

Reply:
I think we need to pass that information from Calcite's LogicalJoin to our JoinNode implementation: https://github.com/apache/pinot/blob/master/pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RelToStageConverter.java#L125